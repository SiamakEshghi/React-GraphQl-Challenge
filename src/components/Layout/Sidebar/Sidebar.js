// @flow
import * as React from 'react';
import { useSelector } from 'react-redux';
import { useTranslation } from 'react-i18next';

import styles from './Sidebar.module.scss';

type SidebarProps = {
  selectFavHandler: (string) => void,
};

const Sidebar = (props: SidebarProps): React.Node => {
  const { selectFavHandler } = props;
  const { favList } = useSelector(({ fav }) => fav);
  const { t } = useTranslation();

  return (
    <ul className={styles.links}>
      <li className={styles.title}>
        <h3>{t('sidebar.title')}</h3>
      </li>
      {favList.map((fav) => {
        const { mbid, name } = fav;
        return (
          <li
            key={mbid}
            className={styles.favLink}
            onClick={() => selectFavHandler(mbid)}
          >
            {name}
          </li>
        );
      })}
    </ul>
  );
};

export default Sidebar;
