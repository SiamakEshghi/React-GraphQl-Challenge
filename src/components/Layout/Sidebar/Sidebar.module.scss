@import '../../../styles/variables/colors.scss';

.links {
  margin-top: 1rem;
  color: $white;
  display: grid;
  grid-template-rows: repeat(auto-fill, 2rem);
  justify-items: center;
  height: 100%;
  list-style: none;
}

.title {
  text-transform: uppercase;
}
.favLink {
  font-size: 1.2rem;
  font-weight: 500;
  text-decoration: none;
  cursor: pointer;
  &:hover {
    color: $tertiaryColor;
  }
}
