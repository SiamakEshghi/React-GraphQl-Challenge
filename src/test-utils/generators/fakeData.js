export const createArtist = (mbid) => ({
  mbid,
  name: 'name-1',
  begin: '1-1-1997',
  end: '',
  gender: 'female',
  rating: null,
  country: '',
});

export const createFav = (mbid) => ({
  mbid,
  name: `name-${mbid}`,
});
