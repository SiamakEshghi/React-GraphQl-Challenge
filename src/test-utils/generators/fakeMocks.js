import { GET_ARTIST, GET_ALBUM, GET_ARTISTS_LIST } from '../../graphQL/queries';

export const createArtistMock = (artistId) => {
  const data = {
    lookup: {
      artist: {
        mbid: 'artistId',
        name: 'artistName',
        gender: 'female',
        country: 'CA',
        rating: 4,
        fanArt: {
          thumbnails: [
            {
              url: 'url',
            },
          ],
        },
        lifeSpan: {
          begin: '2-2-1997',
          end: '1-1-20',
        },
        releaseGroups: {
          edges: [
            {
              node: {
                title: 'album1',
                mbid: 'album1Mbid',
              },
            },
            {
              node: {
                title: 'album2',
                mbid: 'album2Mbid',
              },
            },
          ],
        },
      },
    },
  };

  const mocks = [
    {
      request: {
        query: GET_ARTIST,
        variables: {
          mbid: artistId,
        },
      },
      result: {
        data,
      },
    },
  ];

  return mocks;
};

export const createAlbumMock = (albumMbid) => {
  const data = {
    lookup: {
      releaseGroups: {
        title: 'albumTitle',
        firstReleaseDate: '1-1-2000',
        rating: 3,
        discogs: {
          forSaleCount: 2,
          lowestPrice: 1,
          year: 2002,
        },
      },
    },
  };

  const mocks = [
    {
      request: {
        query: GET_ALBUM,
        variables: {
          mbid: albumMbid,
        },
      },
      result: {
        data,
      },
    },
  ];

  return mocks;
};

export const createArtistsListMock = (name) => {
  const endCursor = 'cursor2';

  const data = {
    search: {
      artists: {
        edges: [
          {
            cursor: 'cursor1',
            node: {
              name: 'name1',
              mbid: 'mbid1',
              fanArt: {
                thumbnails: [
                  {
                    url: 'url',
                  },
                ],
              },
            },
          },
          {
            cursor: 'cursor2',
            node: {
              name: 'name2',
              mbid: 'mbid2',
              fanArt: {
                thumbnails: [
                  {
                    url: 'url',
                  },
                ],
              },
            },
          },
        ],
      },
      pageInfo: {
        endCursor,
      },
    },
  };

  const mocks = [
    {
      request: {
        query: GET_ARTISTS_LIST,
        variables: {
          name,
          endCursor,
        },
      },
      result: {
        data,
      },
    },
  ];

  return mocks;
};
