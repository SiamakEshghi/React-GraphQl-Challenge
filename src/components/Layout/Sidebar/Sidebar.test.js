import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import { useSelector } from 'react-redux';

import Sidebar from './Sidebar';
import { createFav } from '../../../test-utils/generators/fakeData';

const mockEvent = jest.fn();

// mocking useSelector
jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
}));

const defaultProps = {
  selectFavHandler: mockEvent,
};

const setup = (partialState, partialProps) => {
  useSelector.mockImplementation((selector) => selector(partialState));
  return render(<Sidebar {...partialProps} />);
};

describe('Sidebar', () => {
  afterEach(() => {
    mockEvent.mockClear();
    jest.clearAllMocks();
  });

  afterAll(() => {
    jest.restoreAllMocks();
  });

  test('renders without error', () => {
    const partialState = {
      fav: {
        favList: [],
      },
    };

    const { container } = setup(partialState);

    expect(container).toMatchSnapshot();
  });

  test('render correct amount of li based on length of favList. Amount should be favList.length + 1 because first li is the title', () => {
    const favObj1 = createFav('mbid-1');
    const favObj2 = createFav('mbid-2');
    const favObj3 = createFav('mbid-3');

    const partialState = {
      fav: {
        favList: [favObj1, favObj2, favObj3],
      },
    };

    const { container } = setup(partialState);
    const favLiElements = container.querySelectorAll('li');

    expect(favLiElements.length).toBe(4);
  });

  test('click on link call function with fav mbid property', () => {
    const favObj1 = createFav('mbid-1');

    const partialState = {
      fav: {
        favList: [favObj1],
      },
    };

    setup(partialState, defaultProps);

    const liElement = screen.getByText(favObj1.name);

    fireEvent.click(liElement);

    expect(mockEvent).toBeCalledWith(favObj1.mbid);
  });
});
