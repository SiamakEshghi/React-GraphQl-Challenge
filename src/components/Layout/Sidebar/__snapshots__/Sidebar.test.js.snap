// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar renders without error 1`] = `
<div>
  <ul
    class="links"
  >
    <li
      class="title"
    >
      <h3>
        sidebar.title
      </h3>
    </li>
  </ul>
</div>
`;
